framework/data/BaseDataProvider.php

```diff
@@ -44,6 +50,20 @@
     private $_keys;
     private $_models;
     private $_totalCount;
+    
+    /**
+     * @inheritdoc
+     */
+    public function init()
+    {
+        parent::init();
+        if (!isset($this->id)) {
+            if (self::$counter > 0) {
```

Comment:
I agree, but i'd prefer to make that a separate issue.
Otherwise PRs for small fixes / improvements get bogged down due to requiring framework-wide refactoring.